.werft/delete-preview-environments/delete-preview-environments-cron.ts

```diff
@@ -0,0 +1,97 @@
+import { Werft } from '../util/werft';
+import * as Tracing from '../observability/tracing';
+import { SpanStatusCode } from '@opentelemetry/api';
+import { wipePreviewEnvironmentAndNamespace, helmInstallName, listAllPreviewNamespaces } from '../util/kubectl';
+import { exec } from '../util/shell';
+
+// Will be set once tracing has been initialized
+let werft: Werft
+
+Tracing.initialize()
+    .then(() => {
+        werft = new Werft("delete-preview-environment-cron")
+    })
+    .then(() => deletePreviewEnvironments())
+    .then(() => werft.endAllSpans())
+    .catch((err) => {
+        werft.rootSpan.setStatus({
+            code: SpanStatusCode.ERROR,
+            message: err
+        })
+        werft.endAllSpans()
+    })
+
+async function deletePreviewEnvironments() {
+
+    werft.phase("prep");
+    try {
+        const GCLOUD_SERVICE_ACCOUNT_PATH = "/mnt/secrets/gcp-sa/service-account.json";
+        exec(`gcloud auth activate-service-account --key-file "${GCLOUD_SERVICE_ACCOUNT_PATH}"`);
+        exec('gcloud container clusters get-credentials core-dev --zone europe-west1-b --project gitpod-core-dev');
+    } catch (err) {
+        werft.fail("prep", err)
+    }
+    werft.done("prep")
+
+    werft.phase("Fetching branches");
+    const branches = getAllBranches();
+    werft.done("Fetching branches");
+
+    werft.phase("Fetching previews");
+    let previews
+    try {
+        previews = listAllPreviewNamespaces({});
+    } catch (err) {
+        werft.fail("Fetching previews", err)
+    }
+    werft.done("Fetching previews");
+
+    werft.phase("Mapping previews => branches")
+    var previewBranchMap = new Map<string, string>()
+    previews.forEach(preview => {
+        branches.forEach(branch => {
```

Comment:
It not too big a deal as the we're not dealing with a ton of items in the list, but this means you'll be iterating over all the branches for every preview environment. What you could do instead is to create a Set and use the `has` method as checking for set membership is supposed to be fast ☺️ Unfortunately ECMAScript doesn't include any of the Set operations you'd expect (union, difference, etc.) so you'll still have to use filter.

But these nested forEach's and the Map could be replaced with the following (I believe)

```ts
const expectedPreviewEnvironmentNamespaces = new Set(['staging-as-tracing-sigterm' ]); // would be something like new Set([...getAllBranches().map(branch => convertToNamespace(branch))])
const actualPreviewEnvironmentNamespaces = ['staging-as-tracing-sigterm', 'staging-ak-jb-backend-debug-flow']; // would be something like listAllPreviewNamespaces({})

const nameSpacesToDelete = actualPreviewEnvironmentNamespaces.filter(ns => !expectedPreviewEnvironmentNamespaces.has(ns))
```